File: server/internal/domain/geojson.go
package domain

import (
	"encoding/json"
)

const (
	geoJSONFeaturePropertyWayName          = "wayName"
	geoJSONFeaturePropertyMunicipalityName = "municipalityName"
)

// GeoJSONGeometry defines the GeoJSON feature interface.
type GeoJSONGeometry interface {
	// GeometryType returns the geometry type name.
	GeometryType() string
}

// GeoJSON defines the GeoJSON interface.
type GeoJSON interface {
	// Type returns the GeoJSON feature type.
	Type() string
}

// GeoJSONGeometryJSON defines the GeoJSON geometry JSON helper structure.
type GeoJSONGeometryJSON struct {
	Type        string `json:"type"`
	Coordinates any    `json:"coordinates"`
}

// GeoJSONGeometryPoint defines the GeoJSON geometry point structure.
type GeoJSONGeometryPoint struct {
	Coordinates [2]float64
}

func (g GeoJSONGeometryPoint) GeometryType() string {
	return "Point"
}

func (g GeoJSONGeometryPoint) MarshalJSON() ([]byte, error) {
	return json.Marshal(GeoJSONGeometryJSON{
		Type:        g.GeometryType(),
		Coordinates: g.Coordinates,
	})
}

// GeoJSONGeometryLineString defines the GeoJSON geometry line string structure.
type GeoJSONGeometryLineString struct {
	Coordinates [][2]float64
}

func (g GeoJSONGeometryLineString) GeometryType() string {
	return "LineString"
}

func (g GeoJSONGeometryLineString) MarshalJSON() ([]byte, error) {
	return json.Marshal(GeoJSONGeometryJSON{
		Type:        g.GeometryType(),
		Coordinates: g.Coordinates,
	})
}

// GeoJSONFeatureProperties defines the GeoJSON feature properties.
type GeoJSONFeatureProperties map[string]any

// WayName returns the name of the closest way to the geometry.
func (p GeoJSONFeatureProperties) WayName() *string {
	if p == nil {
		return nil
	}

	value, ok := p[geoJSONFeaturePropertyWayName]
	if !ok {
		return nil
	}

	valueString, ok := value.(string)
	if !ok {
		return nil
	}

	return &valueString
}

// SetWayName sets the name of the closest way to the geometry.
func (p GeoJSONFeatureProperties) SetWayName(name string) {
	if p == nil {
		return
	}

	p[geoJSONFeaturePropertyWayName] = name
}

// MunicipalityName returns the name of the municipality containing the geometry.
func (p GeoJSONFeatureProperties) MunicipalityName() *string {
	if p == nil {
		return nil
	}

	value, ok := p[geoJSONFeaturePropertyMunicipalityName]
	if !ok {
		return nil
	}

	valueString, ok := value.(string)
	if !ok {
		return nil
	}

	return &valueString
}

// SetMunicipalityName sets the name of the municipality containing the geometry.
func (p GeoJSONFeatureProperties) SetMunicipalityName(name string) {
	if p == nil {
		return
	}

	p[geoJSONFeaturePropertyMunicipalityName] = name
}

// GeoJSONFeature defines the GeoJSON feature structure.
type GeoJSONFeature struct {
	Geometry   GeoJSONGeometry
	Properties GeoJSONFeatureProperties
}

func (f GeoJSONFeature) Type() string {
	return "Feature"
}

// GeoJSONFeatureCollection defines the GeoJSON feature collection structure.
type GeoJSONFeatureCollection struct {
	Features []GeoJSONFeature
}

func (f GeoJSONFeatureCollection) Type() string {
	return "FeatureCollection"
}

File: android/app/src/main/java/com/ecomap/ecomap/domain/GeoJSON.kt
package com.ecomap.ecomap.domain

import android.content.Context
import com.ecomap.ecomap.R

/**
 * Represents the structure of GeoJSON geometry point.
 * @param coordinates GeoJSON geometry coordinates.
 */
data class GeoJSONGeometryPoint(
    var coordinates: DoubleArray = doubleArrayOf(0.0, 0.0), // [longitude, latitude]
)

/**
 * Represents the structure of GeoJSON feature properties.
 * @param wayName Way name of the feature.
 * @param municipalityName Name of the municipality where the feature is located.
 */
data class GeoJSONProperties(
    var wayName: String = "",
    var municipalityName: String = "",
) {
    /**
     * Returns the location name based on the way and municipality name.
     * @return Location name.
     */
    fun getLocationName(context: Context): String {
        var locationName = context.getString(R.string.way_unknown)

        if (wayName.isNotBlank()) {
            locationName = wayName
        }
        if (municipalityName.isNotBlank()) {
            locationName += ", $municipalityName"
        }

        return locationName
    }
}

/**
 * Represents the structure of a GeoJSON feature point.
 * @param geometry GeoJSON geometry.
 * @param properties GeoJSON properties.
 */
data class GeoJSONFeaturePoint(
    var geometry: GeoJSONGeometryPoint = GeoJSONGeometryPoint(),
    var properties: GeoJSONProperties = GeoJSONProperties(),
)
